src/main/utility/counter.rs

```diff
@@ -0,0 +1,178 @@
+    // TODO: convert this so we could count generic types instead of Strings
+    items: BTreeMap<String, u64>,
+}
+
+impl Counter {
+    /// Initializes a new counter map that starts with no keys.
+    fn new() -> Counter {
+        Counter {
+            items: BTreeMap::new(),
+        }
+    }
+
+    /// Increment the counter value for the key given by id.
+    /// Returns the value of the counter after it was incremented.
+    fn add_one(&mut self, id: &str) -> u64 {
+        let count = self.items.entry(id.to_string()).or_insert(0);
+        *count = *count + 1;
+        *count
+    }
+
+    /// Returns the counter value for the key given by id, or 0 if the key has not
+    /// previously been incremented.
+    // Used for unit test
+    #[allow(dead_code)]
+    fn get_value(&mut self, id: &str) -> u64 {
+        match self.items.get(&id.to_string()) {
+            Some(val) => *val,
+            None => 0,
+        }
+    }
+
+    /// Returns a string representation of the counter. The string starts with "Counts:"
+    /// and then lists known keys and values as a space-separated list of "key=value"
+    /// pairs sorted by value with the largest value first.
+    fn to_string(&mut self) -> String {
+        // Sort the counts with the heaviest hitters first
+        let mut item_vec = Vec::from_iter(&self.items);
+        item_vec.sort_by(|&(_, a), &(_, b)| b.cmp(&a));
+
+        // Create a string representation of the counts
+        let mut string = String::from("Counts:");
+        for item in item_vec.iter() {
+            string.push_str(format!(" {}={}", item.0, item.1).as_str());
+        }
+        string
+    }
+}
+
+mod export {
+    use super::*;
+    use std::ffi::CStr;
+    use std::ffi::CString;
+    use std::os::raw::c_char;
+
+    #[no_mangle]
+    pub extern "C" fn counter_new() -> *mut Counter {
+        Box::into_raw(Box::new(Counter::new()))
+    }
+
+    #[no_mangle]
+    pub extern "C" fn counter_free(counter_ptr: *mut Counter) {
+        if counter_ptr.is_null() {
+            return;
+        }
+        unsafe {
+            Box::from_raw(counter_ptr);
+        }
+    }
+
+    #[no_mangle]
+    pub extern "C" fn counter_add_one(counter: *mut Counter, id: *const c_char) -> u64 {
+        assert!(!counter.is_null());
+        assert!(!id.is_null());
+
+        let counter = unsafe { &mut *counter };
+        let id = unsafe { CStr::from_ptr(id) };
+
+        counter.add_one(id.to_str().unwrap())
+    }
+
+    /// Creates a new string representation of the counter, e.g., for logging.
+    /// The returned string must be free'd by passing it to counter_free_string.
+    #[no_mangle]
+    pub extern "C" fn counter_alloc_string(counter: *mut Counter) -> *mut c_char {
+        assert!(!counter.is_null());
+
+        let counter = unsafe { &mut *counter };
+        let string = counter.to_string();
+
+        // Transfer ownership back to caller
+        CString::new(string.as_str()).unwrap().into_raw()
```

Comment:
Optional - I think you can convert String to &str a little more succinctly with `&`. e.g. `&string`

Reply:
Got an error:

```
error[E0277]: the trait bound `std::vec::Vec<u8>: std::convert::From<&std::string::String>` is not satisfied
   --> utility/counter.rs:109:9
    |
109 |         CString::new(&string).unwrap().into_raw()
    |         ^^^^^^^^^^^^ the trait `std::convert::From<&std::string::String>` is not implemented for `std::vec::Vec<u8>`
    | 
   ::: /home/rjansen/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/ffi/c_str.rs:404:19
    |
404 |     pub fn new<T: Into<Vec<u8>>>(t: T) -> Result<CString, NulError> {
    |                   ------------- required by this bound in `std::ffi::CString::new`
    |
    = help: the following implementations were found:
              <std::vec::Vec<T> as std::convert::From<&[T]>>
              <std::vec::Vec<T> as std::convert::From<&mut [T]>>
              <std::vec::Vec<T> as std::convert::From<[T; N]>>
              <std::vec::Vec<T> as std::convert::From<std::borrow::Cow<'a, [T]>>>
            and 6 others
    = note: required because of the requirements on the impl of `std::convert::Into<std::vec::Vec<u8>>` for `&std::string::String`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: could not compile `shadow-rs`.
```

Reply:
Oops, yeah I'd mis-guessed exactly what `CString::new` takes.

Looking at the [docs](https://doc.rust-lang.org/std/ffi/struct.CString.html#creating-a-cstring) though, it looks like we can pass `string` by value, which should "steal/move" the `String`'s heap-allocated string data, instead of having to allocate and copy it from a `&str`. `CString::new(string).unwrap().into_raw()`

Reply:
Yay, that worked!